### PR TITLE
#12: Add ability to buffer chunks to disk instead of memory

### PR DESCRIPTION
**Pull request by jypma**
This allows a (much) creater amount of upload streams to run in parallel.

@joearasin @filosganga Now includes unit tests. Comments on the direction are welcome.

### CODE REVIEW

File: s3-stream/src/main/scala/com/bluelabs/s3stream/DiskBuffer.scala [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+/**
+ * Buffers the complete incoming stream into a file, which can then be read several times afterwards.
+ * 
+ * The stage waits for the incoming stream to complete. After that, it emits a single Chunk item on its output. The Chunk
+ * contains a bytestream source that can be materialized multiple times, and the total size of the file.
+ *  
+ * @param maxMaterializations Number of expected materializations for the completed chunk. After this, the temp file is deleted.
+ * @param maximumSize Maximum size on disk to buffer
+ */
+class DiskBuffer (maxMaterializations: Int, maxSize: Int, tempPath: Option[Path]) extends GraphStage[FlowShape[ByteString, Chunk]] {
```
Comment (filosganga) [inline s3-stream/src/main/scala/com/bluelabs/s3stream/DiskBuffer.scala:32]: Having a `maxMaterializations` parameter is a good idea. I would like to extend it to be either based on materializations or time.

**Comment by joearasin** [comment]
I'm liking where this is going. the Chunk API is pretty cool :)

File: s3-stream/src/main/scala/com/bluelabs/s3stream/DiskBuffer.scala (hunk 1 above)
Comment (jypma) [reply s3-stream/src/main/scala/com/bluelabs/s3stream/DiskBuffer.scala:32]: Yeah I thought about time, but didn't want to bring in a full-on dependency on the actor system scheduler. If you think it's a bit brittle without it, I can probably add it.

Comment (joearasin) [reply s3-stream/src/main/scala/com/bluelabs/s3stream/DiskBuffer.scala:32]: I'm trying to think of the case where maxMaterializations is > 1...

Comment (jypma) [reply s3-stream/src/main/scala/com/bluelabs/s3stream/DiskBuffer.scala:32]: @joearasin 2 is actually the current case: one materialization for calculating the hash, another one for actually sending the request.

Comment (joearasin) [reply s3-stream/src/main/scala/com/bluelabs/s3stream/DiskBuffer.scala:32]: Ah ok :) That makes sense. Would it make sense to move the hashing into this part of the flow, and have hash be part of the chunk?

Comment (jypma) [reply s3-stream/src/main/scala/com/bluelabs/s3stream/DiskBuffer.scala:32]: I did consider that, but at the time felt hashing was more specific to "sending to S3" than it was to "chunking". Having said that, it's probably a fair bit faster to hash "on the way out", and only have to read stuff back once. I'll have a look.

Comment (joearasin) [reply s3-stream/src/main/scala/com/bluelabs/s3stream/DiskBuffer.scala:32]: Yeah -- I'm a bit torn on it for the same reason. It's nice having the clean "sign and issue an AWS request" bit relatively contained. Maybe hashing on the way out would make sense, but doing it with one read by making the request-generation take a Source[Bytestring] that we can read-and-hash at the same time.

Comment (jypma) [reply s3-stream/src/main/scala/com/bluelabs/s3stream/DiskBuffer.scala:32]: We can't read-and-hash at the same time; we need to send the hash to amazon _before_ all of the data, but we don't know the hash until _after_ all the data. That was the reason for buffering in the first place :)

Unless there are obvious performance drops, I'm voting for keeping the current approach in the PR.

**Pull request closed without merging**